source/slang/slang-ir-header-export.cpp

```diff
@@ -0,0 +1,112 @@
+// slang-ir-header-export.cpp
+#include "slang-ir-header-export.h"
+
+#include "slang-ir.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+
+static void addExternCppDecorationTransitively_(IRBuilder & builder, IRInst * inst)
+{
+    // addExternCppDecoration transitively for function params and struct members (nested)
+    
+    if (auto func = as<IRFunc>(inst))
+    {
+        if (!func->findDecoration<IRExternCppDecoration>())
+        {    
+            builder.addExternCppDecoration(func, UnownedStringSlice(""));
+        }
+        for (auto param : func->getFirstBlock()->getParams())
+        {
+            addExternCppDecorationTransitively_(builder, param->getDataType());
+        }
+    }
+    else if (auto type = as<IRType>(inst))
+    {
+        // If it's a basic type, we're done.
+        if (as<IRBasicType>(type) || as<IRVoidType>(type))
+            return;
+        
+        // If it's a struct type, mark for py-export.
+        if (auto structType = as<IRStructType>(type))
+        {
+            IRBuilder structBuilder(structType->getModule());
+
+            // If it already has a header-export decoration, we're done.
+            if (!structType->findDecoration<IRExternCppDecoration>())
+            {    
+                structBuilder.addExternCppDecoration(structType, UnownedStringSlice(""));
+            }
+
+            for (auto field : structType->getFields())
+            {
+                addExternCppDecorationTransitively_(structBuilder, field->getFieldType());
+            }
+            return;
+        }
+        else if (auto arrayType = as<IRArrayType>(type))
```

Comment:
This should be IRArrayTypeBase.

Should also handle IRPtrTypeBase, IRHLSLStructuredBufferTyoeBase etc.